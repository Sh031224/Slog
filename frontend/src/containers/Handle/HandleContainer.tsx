import React, { useCallback, useEffect, useRef, useState } from "react";
import axios from "axios";
import cookies from "js-cookie";
import { NotificationManager } from "react-notifications";
import { observer } from "mobx-react";
import { useRouter } from "next/router";
import useInterval from "react-useinterval";
import { useBeforeunload } from "react-beforeunload";
import Head from "next/head";
import dynamic from "next/dynamic";
import { PostInfoType } from "types/PostType";
import useStore from "lib/hooks/useStore";
import { CreateTempPostResponse, GetPostInfoResponse, GetProfileResponse, UploadFilesResponse } from "types/Response";

const HandlePost = dynamic(() => import("components/Admin/HandlePost"));

const HandleContainer = () => {
  const { store } = useStore();
  const { categoryList, handleCategoryList } = store.CategoryStore;
  const { handleUser, login, handleLoginChange } = store.UserStore;
  const { getPostInfo, uploadFiles, createTempPost, modifyPost, createPost } = store.PostStore;

  const textAreaRef = useRef<HTMLTextAreaElement>(null);

  const [title, setTitle] = useState<string>("");
  const [description, setDescription] = useState<string>("");
  const [content, setContent] = useState<string>("");
  const [categoryIdx, setCategoryIdx] = useState<number>(-1);
  const [thumbnail, setThumbnail] = useState<string>("");
  const [isNew, setIsNew] = useState<boolean>(true);
  const [isSaving, setIsSaving] = useState<boolean>(false);
  const [isTemp, setIsTemp] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(true);
  const [post_info, setPostInfo] = useState<PostInfoType | React.SetStateAction<PostInfoType | any>>({});

  const router = useRouter();
  const { idx } = router.query;

  const createTempPostCallback = useCallback(async () => {
    if (title !== "") {
      await createTempPost(title, description, content, thumbnail, categoryIdx)
        .then((res: CreateTempPostResponse) => {
          router.push(`/handle/[idx]`, `/handle/${res.data.idx}`);
          NotificationManager.success("임시저장 되었습니다.", "Success");
        })
        .catch(() => {
          NotificationManager.error("오류가 발생하였습니다.", "Error");
        });
    } else {
      NotificationManager.warning("제목을 입력하세요.", "Warning");
    }
    setIsSaving(false);
  }, [title, description, content, categoryIdx, thumbnail]);

  const modifyTempPostCallback = useCallback(async () => {
    if (!isNaN(Number(idx))) {
      if (title !== "") {
        const body: {
          title: string;
          description: string | null;
          content: string | null;
          thumbnail: string | null;
          category_idx?: number;
          is_temp?: boolean;
        } = {
          title,
          description,
          content,
          thumbnail,
          category_idx: categoryIdx
        };

        if (content === "") {
          body.content = "임시 저장";
        }
        if (categoryIdx === -1) {
          delete body.category_idx;
        }
        if (description === "") {
          body.description = "임시저장 글입니다.";
        }

        if (thumbnail === "") {
          body.thumbnail = null;
        }

        await modifyPost(Number(idx), body)
          .then(() => {
            if (content === "") {
              setContent("임시 저장");
            }
            if (description === "") {
              setDescription("임시저장 글입니다.");
            }
            if (!isTemp) {
              router.push(`/post/${idx}`);
            }
            NotificationManager.success("저장 되었습니다.", "Success");
          })
          .catch(() => {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          });
      } else {
        NotificationManager.warning("제목을 입력하세요.", "Warning");
      }
      setIsSaving(false);
    }
  }, [title, description, content, categoryIdx, thumbnail]);

  const createPostCallback = useCallback(async () => {
    if (title === "") {
      NotificationManager.warning("제목을 입력하세요.", "Warning");
    } else if (categoryIdx === -1) {
      NotificationManager.warning("카테고리를 입력하세요.", "Warning");
    } else if (content === "") {
      NotificationManager.warning("내용을 입력하세요.", "Warning");
    } else if (description === "") {
      NotificationManager.warning("설명을 입력하세요.", "Warning");
    } else {
      const body: {
        title: string;
        description: string;
        content: string;
        thumbnail: string | null;
        category_idx: number;
      } = {
        title,
        description,
        content,
        thumbnail,
        category_idx: categoryIdx
      };

      if (thumbnail === "") {
        body.thumbnail = null;
      }

      await createPost(body).catch(() => {
        NotificationManager.error("오류가 발생하였습니다.", "Error");
      });
      router.push(`/`);
      NotificationManager.success("저장 되었습니다.", "Success");
    }
    setIsSaving(false);
  }, [title, description, content, categoryIdx, thumbnail]);

  const modifyPostCallback = useCallback(async () => {
    if (title !== "") {
      if (categoryIdx === -1) {
        NotificationManager.warning("카테고리를 입력하세요.", "Warning");
      }
      if (description === "") {
        NotificationManager.warning("설명을 입력하세요.", "Warning");
      }
      if (content === "") {
        NotificationManager.warning("내용을 입력하세요.", "Warning");
      }
      if (content !== "" && categoryIdx !== -1 && description !== "") {
        const body: {
          title: string;
          description: string | null;
          content: string | null;
          thumbnail: string | null;
          category_idx?: number;
          is_temp?: boolean;
        } = {
          title,
          description,
          content,
          thumbnail,
          category_idx: categoryIdx,
          is_temp: false
        };

        if (post_info.is_temp) {
          if (content === "") {
            body.content = null;
          }
          if (categoryIdx === -1) {
            delete body.category_idx;
          }
          if (description === "") {
            body.description = "";
          }
        }

        if (thumbnail === "") {
          body.thumbnail = null;
        }

        await modifyPost(Number(idx), body)
          .then(() => {
            router.push(`/post/${idx}`);
            NotificationManager.success("저장 되었습니다.", "Success");
          })
          .catch(() => {
            NotificationManager.error("오류가 발생하였습니다.", "Error");
          });
      }
    } else {
      NotificationManager.warning("제목을 입력하세요.", "Warning");
    }
    setIsSaving(false);
  }, [title, description, content, categoryIdx, thumbnail, idx]);

  const uploadFilesCallback = useCallback(async (files: File[]) => {
    await uploadFiles(files)
      .then((res: UploadFilesResponse) => {
        setContent((content) => content + `\n![image](${res.data.files[0].replace(" ", "%20")})`);
        NotificationManager.success("사진이 업로드 되었습니다.", "Success");
      })
      .catch(() => {
        NotificationManager.error("오류가 발생하였습니다.", "Error");
      });
  }, []);

  const validateAdmin = useCallback(() => {
    try {
      let token = axios.defaults.headers.common["access_token"];
      if (token) {
        handleLoginChange(true);
        handleUser(token)
          .then((res: GetProfileResponse) => {
            if (!res.data.user.is_admin) {
              router.push("/");
            }
          })
          .catch((err) => {
            if (err.message === "401") {
              cookies.remove("access_token", { path: "/" });
              handleLoginChange(false);
              axios.defaults.headers.common["access_token"] = "";
            }
          });
      } else {
        router.push("/");
        handleLoginChange(false);
      }
    } catch (err) {
      NotificationManager.error("오류가 발생하였습니다.", "Error");
    }
  }, [login]);

  const checkIsNewCallback = useCallback(() => {
    if (idx === "new") {
      setIsNew(true);
      setIsTemp(true);
    } else {
      setIsNew(false);
    }
  }, [idx]);

  const handleCategoryListCallback = useCallback(() => {
    handleCategoryList().catch(() => NotificationManager.error("오류가 발생하였습니다.", "Error"));
  }, []);

  const getPostInfoCallback = useCallback(() => {
    if (!isNew) {
      getPostInfo(Number(idx))
        .then((res: GetPostInfoResponse) => {
          setLoading(false);
          setPostInfo(res.data.post);
          setIsTemp(res.data.post.is_temp);
          setTitle(res.data.post.title);
          if (res.data.post.content !== null) {
            setContent(res.data.post.content);
          }
          if (res.data.post.description !== null) {
            setDescription(res.data.post.description);
          }
          if (res.data.post.fk_category_idx !== null) {
            setCategoryIdx(res.data.post.fk_category_idx);
          }
          if (res.data.post.thumbnail !== null) {
            setThumbnail(res.data.post.thumbnail.replace("https://data.slog.website/public/", ""));
          }
        })
        .catch(() => {
          router.push("/");
          NotificationManager.error("오류가 발생하였습니다.", "Error");
        });
    }
  }, [isNew]);

  const cancelBtn = () => {
    router.back();
  };

  const savePostHandle = useCallback(() => {
    if (!isSaving) {
      setIsSaving(true);
      if (isNew) {
        createPostCallback();
      } else {
        modifyPostCallback();
      }
    }
  }, [createPostCallback, modifyPostCallback, isNew]);

  const tempPostHandle = useCallback(() => {
    if (!isSaving) {
      setIsSaving(true);
      if (isNew) {
        createTempPostCallback();
      } else {
        modifyTempPostCallback();
      }
    }
  }, [createTempPostCallback, modifyTempPostCallback, isNew]);

  useEffect(() => {
    validateAdmin();
  }, [validateAdmin]);

  useEffect(() => {
    checkIsNewCallback();
  }, [checkIsNewCallback]);

  useEffect(() => {
    getPostInfoCallback();
  }, [getPostInfoCallback]);

  useEffect(() => {
    handleCategoryListCallback();
  }, [handleCategoryListCallback]);

  const autoSaveCallback = useCallback(() => {
    if (isTemp) {
      if (title !== "") {
        tempPostHandle();
      }
    }
  }, [title, isTemp, content, thumbnail, categoryIdx, description]);

  useInterval(autoSaveCallback, 250000);

  useBeforeunload(() => "창을 닫으면 모든 변경사항이 사라집니다.");

  return (
    <>
      <Head>
        <title>{"Slog"}</title>
        <meta name="description" content="많은 사람들에게 유용한 정보를 제공하기 위해 제작한 Slog입니다." />
        <meta property="og:title" content="Slog" />
        <meta property="og:description" content="많은 사람들에게 유용한 정보를 제공하기 위해 제작한 Slog입니다." />
        <meta property="og:image" content="https://data.slog.website/public/op_logo.png" />
        <meta name="twitter:card" content="summary_large_image" />
        <meta name="twitter:title" content="Slog" />
        <meta name="twitter:description" content="많은 사람들에게 유용한 정보를 제공하기 위해 제작한 Slog입니다." />
        <meta name="twitter:image" content="https://data.slog.website/public/op_logo.png" />
      </Head>
      {isNew ? (
        <HandlePost
          cancelBtn={cancelBtn}
          categoryList={categoryList}
          loading={false}
          title={title}
          setTitle={setTitle}
          description={description}
          setDescription={setDescription}
          content={content}
          setContent={setContent}
          categoryIdx={categoryIdx}
          setCategoryIdx={setCategoryIdx}
          thumbnail={thumbnail}
          setThumbnail={setThumbnail}
          uploadFilesCallback={uploadFilesCallback}
          textAreaRef={textAreaRef}
          tempPostHandle={tempPostHandle}
          savePostHandle={savePostHandle}
          isTemp={isTemp}
        />
      ) : (
        <HandlePost
          cancelBtn={cancelBtn}
          categoryList={categoryList}
          loading={loading}
          title={title}
          setTitle={setTitle}
          description={description}
          setDescription={setDescription}
          content={content}
          setContent={setContent}
          categoryIdx={categoryIdx}
          setCategoryIdx={setCategoryIdx}
          thumbnail={thumbnail}
          setThumbnail={setThumbnail}
          uploadFilesCallback={uploadFilesCallback}
          textAreaRef={textAreaRef}
          savePostHandle={savePostHandle}
          isTemp={isTemp}
          tempPostHandle={tempPostHandle}
        />
      )}
    </>
  );
};

export default observer(HandleContainer);
