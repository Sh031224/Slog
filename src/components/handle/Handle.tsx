import useHandlePost from "hooks/handle/useHandlePost";

import styled from "styled-components";
import HandleContent from "./content/HandleContent";
import HandleFooter from "./footer/HandleFooter";
import HandleHeader from "./header/HandleHeader";
import HandleUtil from "./util/HandleUtil";

const Handle: React.FC = () => {
  const {
    value,
    handleValue,
    categories,
    isPreviewTab,
    setIsPreviewTab,
    handleContentTab,
    handleUpload,
    textareaRef,
    isNew,
    save,
    tempSave
  } = useHandlePost();

  return (
    <HandleWrapper>
      <HandleBox>
        <HandleHeader value={value} handleValue={handleValue} />
        <HandleUtil value={value} handleValue={handleValue} categories={categories} />
        <HandleContent
          value={value}
          handleValue={handleValue}
          isPreviewTab={isPreviewTab}
          setIsPreviewTab={setIsPreviewTab}
          textareaRef={textareaRef}
          handleContentTab={handleContentTab}
          handleUpload={handleUpload}
        />
        <HandleFooter isNew={isNew} isTemp={value.is_temp} save={save} tempSave={tempSave} />
      </HandleBox>
    </HandleWrapper>
  );
};

const HandleWrapper = styled.main`
  width: 100%;
  display: flex;
  justify-content: center;
`;

const HandleBox = styled.div`
  max-width: 900px;
  position: relative;
  width: 100%;
  padding: 0 1rem;
`;

export default Handle;
