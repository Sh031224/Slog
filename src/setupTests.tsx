import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { lightTheme } from "style/theme";
import { store } from "stores";
import { Provider } from "react-redux";
import { NotificationContainer } from "react-notifications";

const Providers = ({ children }) => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={lightTheme}>
        <NotificationContainer />
        {children}
      </ThemeProvider>
    </Provider>
  );
};

const customRender = (ui, options = {}) => render(ui, { wrapper: Providers, ...options });

// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

(window as any).scrollTo = () => {};

export { customRender as render };
