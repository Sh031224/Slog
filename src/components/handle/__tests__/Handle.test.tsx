import { render } from "setupTests";
import Handle from "../Handle";

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/handle/1",
      pathname: "",
      query: { idx: 1 },
      asPath: "",
      push: jest.fn()
    };
  }
}));

describe("Handle", () => {
  const setup = (props = {}) => {
    return render(<Handle {...props} />);
  };

  it("renders correctly", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });
});
