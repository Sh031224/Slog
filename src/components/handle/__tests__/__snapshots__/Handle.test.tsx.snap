// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle renders correctly 1`] = `
.c17 {
  font-size: 1.125rem;
  color: #222426;
  line-height: 1.7;
  -webkit-letter-spacing: -0.004em;
  -moz-letter-spacing: -0.004em;
  -ms-letter-spacing: -0.004em;
  letter-spacing: -0.004em;
  word-break: keep-all;
  overflow-wrap: break-word;
}

.c17 p + a,
.c17 a {
  color: #5373f8;
}

.c17 p + a:hover,
.c17 a:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c17 p + h1,
.c17 p + h2,
.c17 p + h3,
.c17 p + h4 {
  margin-top: 2.5rem;
}

.c17 h1,
.c17 h2,
.c17 h3,
.c17 h4 {
  line-height: 1.5;
  margin-bottom: 1rem;
}

.c17 h1 {
  font-size: 2.5rem;
}

.c17 h2 {
  font-size: 2rem;
}

.c17 h3 {
  font-size: 1.5rem;
}

.c17 h4 {
  font-size: 1.125rem;
}

.c17 hr {
  height: 1px;
  width: 100%;
  margin-top: 2rem;
  margin-bottom: 2rem;
  border-width: initial;
  border-style: none;
  border-color: initial;
  border-image: initial;
  background: #dedede;
}

.c17 blockquote {
  margin-top: 2rem;
  margin-bottom: 2rem;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  margin-left: 0px;
  margin-right: 0px;
  color: rgb(33,37,41);
  border-left: #5373f8 4px solid;
  background-color: rgb(241,243,245);
  padding: 1rem 1rem 1rem 2rem;
}

.c17 blockquote :first-child {
  margin-top: 0px;
}

.c17 blockquote :last-child {
  margin-bottom: 0px;
}

.c17 pre > div {
  font-family: "Fira Mono",source-code-pro,Menlo,Monaco,Consolas,"Courier New",monospace;
  overflow-x: auto;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  border-radius: 5px;
  line-height: 1.5 !important;
  overflow-x: auto !important;
  -webkit-letter-spacing: 0px !important;
  -moz-letter-spacing: 0px !important;
  -ms-letter-spacing: 0px !important;
  letter-spacing: 0px !important;
  padding: 1rem !important;
  border-radius: 5px !important;
  font-size: 0.875rem;
}

.c17 pre > div > code {
  background: none !important;
}

.c17:not(pre) > code[class*="language-"],
.c17 pre[class*="language-"] {
  background: none !important;
}

.c17 code {
  font-family: "Fira Mono",source-code-pro,Menlo,Monaco,Consolas,"Courier New",monospace;
}

.c17 p > code {
  background: #1b1f231c;
  padding: 0.2em 0.4em;
  border-radius: 5px;
  font-size: 85%;
}

.c17 img {
  max-width: 100%;
  height: auto;
  display: block;
  margin-top: 1.5rem;
  margin-bottom: 1.5rem;
}

.c17 p > img {
  display: block;
  max-width: 100%;
  margin: 3rem auto;
}

.c17 table {
  border-collapse: collapse;
  display: block;
  width: 100%;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 100%;
  overflow: auto;
}

.c17 table tr:nth-child(2n) {
  background-color: #f6f8fa;
}

.c17 table tr {
  background-color: #ffff;
  border-top: 1px solid #c6cbd1;
}

.c17 table td,
.c17 table th {
  padding: 6px 13px;
  border: 1px solid #dfe2e5;
}

.c7 {
  margin-top: 2rem;
}

.c7 input[type="file"] {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  border: 0;
}

.c8 {
  width: 100%;
  border-radius: 5px;
  border: #e1e4e8 1px solid;
  position: relative;
}

.c9 {
  display: table;
  width: 100%;
  border-bottom: #e1e4e8 1px solid;
  background-color: #fafbfc;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  position: relative;
  padding: 0.25rem 0.5rem;
}

.c10 {
  margin: calc(-0.25rem - 1px) 0 calc(-0.25rem - 1.5px) calc(-0.5rem - 1.5px);
  cursor: pointer;
  display: inline-block;
}

.c11 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.8rem 1.2rem;
  color: #222426;
  line-height: 23px;
  border: none;
  font-size: 0.9rem;
  cursor: pointer;
  background-color: #ffff;
  border: 1px solid #e1e4e8;
  border-bottom: none;
}

.c11:first-child {
  border-top-left-radius: 5px;
}

.c13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.8rem 1.2rem;
  color: #222426;
  line-height: 23px;
  border: none;
  font-size: 0.9rem;
  cursor: pointer;
  background-color: transparent;
}

.c13:first-child {
  border-top-left-radius: 5px;
}

.c12 {
  font-size: 1.1rem;
  color: #a0a0a0;
  margin-right: 0.8rem;
}

.c14 {
  font-size: 1.1rem;
  color: #a0a0a0;
  margin-right: 0.8rem;
}

.c15 {
  display: inline-block;
  width: calc(100% - 20rem);
  text-align: center;
  font-size: 1.2rem;
  color: #464646;
}

.c15 svg {
  cursor: pointer;
}

.c18 {
  font-family: NotoSansKR;
  padding: 1rem;
  font-size: 1rem;
  border: none;
  display: block;
  resize: none;
  width: 100%;
  height: 40rem;
  overflow: auto;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
}

.c18:focus {
  outline: none;
}

.c16 {
  width: 100%;
  border-bottom-left-radius: 4px;
  border-bottom-right-radius: 4px;
  padding: 0.3rem 1rem;
  height: 40rem;
  overflow: scroll;
  display: none;
}

.c19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-top: 1rem;
}

.c20 {
  border: none;
  padding: 0.5rem 1rem;
  margin-left: 1.5rem;
  box-shadow: 1px 1px 10px rgba(0,0,0,0.096);
  cursor: pointer;
  border-radius: 5px;
  background-color: #ffff;
}

.c20:last-child {
  background-color: #5373f8;
  color: #fff;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 3rem;
}

.c3 {
  border: #e1e4e8 1px solid;
  border-radius: 5px;
  padding: 0.75rem;
  outline: none;
  margin-bottom: 1rem;
}

.c3:focus {
  border: #c0c0c0 2px solid;
}

.c3:first-child {
  font-size: 1.3rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 1rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 {
  padding: 0.4rem;
  font-size: 0.8rem;
  outline: none;
  width: 65%;
  border: #e1e4e8 1px solid;
  border-radius: 5px;
  outline: none;
}

.c5:focus {
  border: #c0c0c0 2px solid;
}

.c6 {
  width: 35%;
  border: #e1e4e8 1px solid;
  border-radius: 5px;
  padding: 0.4rem;
  font-size: 0.8rem;
  outline: none;
  margin-left: 1rem;
  color: black;
}

.c6:invalid {
  color: #777;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  max-width: 900px;
  position: relative;
  width: 100%;
  padding: 0 1rem;
}

@media only screen and (max-width:767px) {
  .c17 {
    font-size: 1rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 p + h1,
  .c17 p + h2,
  .c17 p + h3,
  .c17 p + h4 {
    margin-top: 2rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 h1,
  .c17 h2,
  .c17 h3,
  .c17 h4 {
    margin-bottom: 0.75rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 h1 {
    font-size: 2.25rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 h2 {
    font-size: 1.75rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 h3 {
    font-size: 1.25rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 h4 {
    font-size: 1rem;
  }
}

@media only screen and (max-width:767px) {
  .c17 pre > div {
    font-size: 0.75rem;
  }
}

@media only screen and (max-width:500px) {
  .c7 .upload-image {
    width: 15rem !important;
  }

  .c7 .upload-image p {
    font-size: 0.7rem !important;
  }
}

@media only screen and (max-width:500px) {
  .c11 {
    font-size: 0.7rem;
    padding: 0.5rem 0.9rem;
  }
}

@media only screen and (max-width:500px) {
  .c13 {
    font-size: 0.7rem;
    padding: 0.5rem 0.9rem;
  }
}

@media only screen and (max-width:500px) {
  .c12 {
    font-size: 0.8rem;
    margin-right: 0.5rem;
  }
}

@media only screen and (max-width:500px) {
  .c14 {
    font-size: 0.8rem;
    margin-right: 0.5rem;
  }
}

@media only screen and (max-width:500px) {
  .c15 {
    width: calc(100% - 13rem);
    font-size: 1rem;
  }
}

<div>
  <div
    class="notification-container notification-container-empty"
  >
    <div />
  </div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <section
        class="c2"
      >
        <input
          class="c3"
          maxlength="255"
          name="title"
          placeholder="제목을 입력하세요."
          type="text"
          value=""
        />
        <input
          class="c3"
          maxlength="255"
          name="description"
          placeholder="설명을 입력하세요."
          type="text"
          value=""
        />
      </section>
      <section
        class="c4"
      >
        <input
          class="c5"
          name="thumbnail"
          placeholder="미리보기 이미지를 넣어주세요."
          type="text"
          value=""
        />
        <select
          class="c6"
          name="category_idx"
          required=""
        >
          <option
            value=""
          >
            카테고리를 선택해주세요.
          </option>
        </select>
      </section>
      <section
        class="c7"
      >
        <div
          class="c8"
        >
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <button
                aria-label="change-tab"
                class="c11"
              >
                <svg
                  class="c12"
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polyline
                    points="16 18 22 12 16 6"
                  />
                  <polyline
                    points="8 6 2 12 8 18"
                  />
                </svg>
                Edit Content
              </button>
              <button
                aria-label="change-preview"
                class="c13"
              >
                <svg
                  class="c14"
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    r="3"
                  />
                </svg>
                Preview
              </button>
            </div>
            <div
              class="c15"
            >
              <label
                for="upload"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 632H136v-39.9l138.5-164.3 150.1 178L658.1 489 888 761.6V792zm0-129.8L664.2 396.8c-3.2-3.8-9-3.8-12.2 0L424.6 666.4l-144-170.7c-3.2-3.8-9-3.8-12.2 0L136 652.7V232h752v430.2zM304 456a88 88 0 1 0 0-176 88 88 0 0 0 0 176zm0-116c15.5 0 28 12.5 28 28s-12.5 28-28 28-28-12.5-28-28 12.5-28 28-28z"
                  />
                </svg>
              </label>
              <input
                accept="png, gif, jpeg, jpg"
                id="upload"
                type="file"
              />
            </div>
          </div>
          <div
            class="c16"
          >
            <div
              class="c17"
            >
              <span>
                

              </span>
            </div>
          </div>
          <textarea
            class="c18"
            name="content"
          />
        </div>
      </section>
      <section
        class="c19"
      >
        <button
          aria-label="cancel-handle"
          class="c20"
        >
          취소
        </button>
        <button
          aria-label="temp-handle"
          class="c20"
        >
          임시 저장
        </button>
        <button
          aria-label="save-handle"
          class="c20"
        >
          저장
        </button>
      </section>
    </div>
  </main>
</div>
`;
